src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -826,19 +890,10 @@ fn traits_implemented_by(cx: &mut DocContext<'_>, type_: DefId, module: DefId) -
                 "comparing type {} with kind {:?} against type {:?}",
                 impl_type,
                 impl_type.kind(),
-                type_
+                ty
             );
             // Fast path: if this is a primitive simple `==` will work
-            let saw_impl = impl_type == ty
-                || match impl_type.kind() {
-                    // Check if these are the same def_id
-                    ty::Adt(def, _) => {
-                        debug!("adt def_id: {:?}", def.did);
-                        def.did == type_
-                    }
-                    ty::Foreign(def_id) => *def_id == type_,
-                    _ => false,
-                };
+            let saw_impl = impl_type == ty;
```

Comment:
My understanding is that the `match` that was here before is not necessary, since `ty::Adt`'s `eq` implementation does basically compare the `did`:

https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_middle/ty/adt.rs.html#120-122

Similarly for `Foreign`, it will just compare the `def_id` inside.

Reply:
Can you move this change into its own PR?

Reply:
@camelid this is necessary to make this function work with a `ty` input, the `match` is only useful if the function accepts DefId. I don't understand how it can be its own PR.

Reply:
I'm not sure why it's necessary; the `match` doesn't seem to use the function's DefId parameter.

But either way, I meant can you open another that makes just this change, which this PR can then be rebased upon once the new one is merged? It seems like a cleanup that should be made independently of this feature PR.

Reply:
The match is there because we are trying to see if the `ty` that we have found is equal to the `DefId` that we have, hence we have a match that is like this:

```
match impl_type.kind() {
                    // Check if these are the same def_id
                    ty::Adt(def, _) => {
                        debug!("adt def_id: {:?}", def.did);
                        def.did == type_
                    }
                    ty::Foreign(def_id) => *def_id == type_,
                    _ => false,
                };
```

The `type_` here is the `DefId` we receive in the arguments of the function, however since I have removed that `DefId` and turned it into a `ty`, and `ty` handles these cases of `Foreign` and `Adt` properly, a simple `==` is enough here.

Reply:
Now, I'm not sure if the `tcx.type_of` previously used to get a `ty` instance out of the `type_ : DefId` (`let ty = tcx.type_of(type_);`) would achieve the same thing: is that `ty` also fully informative so that the equal sign is enough? Perhaps... in that case we can have a separate PR, but I'm not sure if that's the case.

Reply:
this seems okay to me